File: textdrive-c/train.c
/*
 * Q学習でゲームを学習するプログラム
 *
 * 学習の流れ:
 *   1. 状態を見る
 *   2. 行動を選ぶ
 *   3. 行動する
 *   4. 報酬をもらう
 *   5. Q値を更新する
 *   6. 1に戻る
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "game.h"
#include "qlearning.h"

#define NUM_EPISODES 50000   /* 学習回数 */
#define MAX_STEPS    10000   /* 1ゲームの最大ステップ */

int main(void) {
    srand((unsigned)time(NULL));

    /* エージェントを初期化 */
    Agent agent;
    agent_init(&agent);

    printf("=== Q学習スタート ===\n\n");

    /* 直近500ゲームの平均を計算するための変数 */
    int recent_scores[500];
    int recent_idx = 0;
    long recent_sum = 0;
    int recent_count = 0;

    /* 学習ループ */
    for (int episode = 1; episode <= NUM_EPISODES; episode++) {

        /* ゲームを初期化 */
        Game game;
        initGame(&game);

        /* 1ゲームをプレイ */
        for (int step = 0; step < MAX_STEPS; step++) {

            /* 1. 状態を見る */
            int state = get_state(&game);

            /* 2. 行動を選ぶ */
            int action = choose_action(&agent, state);

            /* 3. 行動する */
            do_action(&game, action);
            scrollCourse(&game);

            /* 4. 報酬をもらう */
            double reward = get_reward(&game);

            /* 5. 次の状態を見る */
            int next_state = get_state(&game);

            /* 6. Q値を更新する */
            update_Q(&agent, state, action, reward, next_state);

            /* 衝突したら終了 */
            if (hasCollision(&game)) {
                break;
            }
        }

        /* スコアを記録 */
        if (game.distance > agent.best_score) {
            agent.best_score = game.distance;
        }
        agent.episodes++;

        /* 直近平均を更新 */
        if (recent_count == 500) {
            recent_sum -= recent_scores[recent_idx];
        } else {
            recent_count++;
        }
        recent_scores[recent_idx] = game.distance;
        recent_sum += game.distance;
        recent_idx = (recent_idx + 1) % 500;

        /* 探索率を減らす */
        agent_decay_epsilon(&agent);

        /* 進捗を表示 */
        if (episode % 500 == 0) {
            double avg = (double)recent_sum / recent_count;
            printf("Episode %5d | Best: %5d | Avg: %7.1f | ε: %.3f\n",
                   episode, agent.best_score, avg, agent.epsilon);
        }
    }

    printf("\n=== 学習完了 ===\n");
    printf("最高スコア: %d\n", agent.best_score);

    /* 保存 */
    agent_save(&agent, "qtable.bin");
    printf("データを保存しました\n");

    return 0;
}

File: textdrive-rust/src/bin/train.rs
use textdrive::game::Game;
use textdrive::qlearning::{choose_action, do_action, get_reward, get_state, update_q, Agent};

const NUM_EPISODES: usize = 50000;
const MAX_STEPS: usize = 10000;
const REPORT_INTERVAL: usize = 500;
const RECENT_WINDOW_SIZE: usize = 500;

fn main() {
    let mut agent = Agent::new();
    let mut stats = TrainingStats::new(RECENT_WINDOW_SIZE);

    println!("=== Q学習スタート ===\n");

    for episode in 1..=NUM_EPISODES {
        let final_distance = run_episode(&mut agent);

        stats.update(final_distance, &agent);
        agent.decay_epsilon();

        if episode % REPORT_INTERVAL == 0 {
            stats.print_progress_with_agent(episode, &agent);
        }
    }

    println!("\n=== 学習完了 ===");
    println!("最高スコア: {}", agent.best_score);

    agent
        .save("qtable.bin")
        .expect("データの保存に失敗しました");
    println!("データを保存しました");
}

fn run_episode(agent: &mut Agent) -> i32 {
    let mut game = Game::new();

    for _step in 0..MAX_STEPS {
        let state = get_state(&game);
        let action = choose_action(agent, state);

        do_action(&mut game, action);
        game.scroll_course();

        let reward = get_reward(&game);
        let next_state = get_state(&game);

        update_q(agent, state, action, reward, next_state);

        if game.has_collision() {
            break;
        }
    }

    if game.distance > agent.best_score {
        agent.best_score = game.distance;
    }
    agent.episodes += 1;

    game.distance
}

struct TrainingStats {
    recent_scores: Vec<i32>,
    recent_idx: usize,
    recent_sum: i64,
    recent_count: usize,
    window_size: usize,
}

impl TrainingStats {
    fn new(window_size: usize) -> Self {
        Self {
            recent_scores: vec![0; window_size],
            recent_idx: 0,
            recent_sum: 0,
            recent_count: 0,
            window_size,
        }
    }

    fn update(&mut self, score: i32, _agent: &Agent) {
        if self.recent_count == self.window_size {
            self.recent_sum -= self.recent_scores[self.recent_idx] as i64;
        } else {
            self.recent_count += 1;
        }

        self.recent_scores[self.recent_idx] = score;
        self.recent_sum += score as i64;
        self.recent_idx = (self.recent_idx + 1) % self.window_size;
    }

    fn average(&self) -> f64 {
        if self.recent_count == 0 {
            0.0
        } else {
            self.recent_sum as f64 / self.recent_count as f64
        }
    }

    fn print_progress_with_agent(&self, episode: usize, agent: &Agent) {
        println!(
            "Episode {:5} | Best: {:5} | Avg: {:7.1} | ε: {:.3}",
            episode,
            agent.best_score,
            self.average(),
            agent.epsilon
        );
    }
}
